### PR TITLE
feat: customize Typography.Paragraph copyable and editable

### PR DESCRIPTION
**Pull request by llwslc**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link

resolve https://github.com/ant-design/ant-design/issues/25947
resolve https://github.com/ant-design/ant-design/issues/25994
resolve https://github.com/ant-design/ant-design/issues/19562

### 💡 Background and solution

- copyable
  - add `icon` prop
  - add `tooltips` prop
- editable
  - add `icon` prop
  - add `tooltip` prop

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |  Typography emhance`copyable` and `editable` props. |
| 🇨🇳 Chinese |  Typography 增强 `copyable` 和 `editable` 属性。  |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

-----
[View rendered components/typography/demo/interactive.md](https://github.com/llwslc/ant-design/blob/customize-copyable/components/typography/demo/interactive.md)
[View rendered components/typography/index.en-US.md](https://github.com/llwslc/ant-design/blob/customize-copyable/components/typography/index.en-US.md)
[View rendered components/typography/index.zh-CN.md](https://github.com/llwslc/ant-design/blob/customize-copyable/components/typography/index.zh-CN.md)

### CODE REVIEW

**Comment by afc163** [comment]
image

File: components/typography/Base.tsx [hunk 1]
```diff
@@ -28,8 +28,8 @@ const isTextOverflowSupport = isStyleSupport('textOverflow');
 interface CopyConfig {
   text?: string;
   onCopy?: () => void;
-  icon?: React.ReactNode;
-  tooltips?: [React.ReactNode, React.ReactNode];
+  icon?: React.ReactNode | Array<React.ReactNode>;
```
Comment (07akioni) [inline components/typography/Base.tsx:31]: `React.ReactNode` already contains `Array<React.ReactNode>`

File: components/typography/Base.tsx [hunk 2]
```diff
@@ -28,8 +28,8 @@ const isTextOverflowSupport = isStyleSupport('textOverflow');
...
+  tooltips?: boolean | Array<React.ReactNode>;
 }
```
Comment (07akioni) [inline components/typography/Base.tsx:33]: the same

File: components/typography/Base.tsx [hunk 3]
```diff
@@ -383,11 +383,19 @@ class Base extends React.Component<InternalBlockProps, BaseState> {
...
-      : (copyable as CopyConfig).tooltips?.[0] || this.copyStr;
+    const { tooltips } = copyable as CopyConfig;
+    let title: string | React.ReactNode = '';
+    if (Array.isArray(tooltips)) {
+      const tooltipsNode = toArray(tooltips);
+      title = copied ? tooltipsNode[1] || this.copiedStr : tooltipsNode[0] || this.copyStr;
+    } else if (tooltips === undefined || tooltips) {
+      title = copied ? this.copiedStr : this.copyStr;
+    } else {
+      title = '';
```
Comment (07akioni) [inline components/typography/Base.tsx:394]: not need for re-assign

**Comment by afc163** [comment]
Could you handle https://github.com/ant-design/ant-design/issues/25994 too?

**Comment by llwslc** [comment]
> Could you handle #25994 too?

Copy that.

**Comment by afc163** [comment]
rebase 一下最新的 feature 分支。

File: components/typography/demo/interactive.md [hunk 4]
```diff
@@ -30,30 +32,54 @@ class Demo extends React.Component {
...
+  onHideTooltipStrChange = hideTooltipStr => {
+    this.setState({ hideTooltipStr });
+  };
+
   onLengthLimitedStrChange = lengthLimitedStr => {
     this.setState({ lengthLimitedStr });
   };
 
   render() {
-    const { lengthLimitedStr } = this.state;
```
Comment (afc163) [inline components/typography/demo/interactive.md:38]: state 统一提到前面解构吧。

Comment (afc163) [reply components/typography/demo/interactive.md:38]: 干脆例子也改成 function component 好了。

File: components/typography/index.en-US.md [hunk 5]
```diff
@@ -20,10 +20,10 @@ Basic text writing, including headings, body text, lists, and more.
 | Property | Description | Type | Default | Version |
 | --- | --- | --- | --- | --- |
 | code | Code style | boolean | false |  |
-| copyable | Whether to be copyable, customize it via setting an object | boolean \| { text: string, onCopy: function, icon: ReactNode, tooltips: \[ReactNode, ReactNode\] } | false | icon and tooltips: 4.4.0 |
+| copyable | Whether to be copyable, customize it via setting an object | boolean \| { text: string, onCopy: function, icon: ReactNode, tooltips: boolean \| ReactNode } | false | `icon` and `tooltips`: 4.4.0 `icon: ReactNode[]` and `tooltips: false`: 4.6.0 |
```
Comment (xrkffgg) [inline components/typography/index.en-US.md:23]: 没明白 icon 和 tooltips 列出 2 个版本是啥意思？

Comment (xrkffgg) [reply components/typography/index.en-US.md:23]: 版本中属性不加 ` ` 了，后续会逐步统一

Comment (llwslc) [reply components/typography/index.en-US.md:23]: 哦, 那 4.4.0 不支持修改 copied icon 该怎么表达呢?

Comment (xrkffgg) [reply components/typography/index.en-US.md:23]: icon and tooltips: 4.6.0 直接这样吧

Comment (afc163) [reply components/typography/index.en-US.md:23]: 单独在后面写个版本兼容表格，然后这里留一个 hash link 过来好了，单元格里太挤了。

Comment (llwslc) [reply components/typography/index.en-US.md]: 我没有找到先例, 可以告诉我照哪个写一下吗?

Comment (afc163) [reply components/typography/index.en-US.md]: 例如 rules 的写法：https://ant.design/components/form-cn/#Form.Item

Comment (llwslc) [reply components/typography/index.en-US.md]: 看下这样可以吗?

File: components/typography/index.zh-CN.md [hunk 6]
```diff
@@ -53,17 +53,27 @@ cover: https://gw.alipayobjects.com/zos/alicdn/GOM1KQ24O/Typography.svg
 | 参数 | 说明 | 类型 | 默认值 | 版本 |
 | --- | --- | --- | --- | --- |
 | code | 添加代码样式 | boolean | false |  |
-| copyable | 是否可拷贝，为对象时可进行各种自定义 | boolean \| { text: string, onCopy: function, icon: ReactNode, tooltips: \[ReactNode, ReactNode\] } | false | `icon` 和 `tooltips` 在 `4.4.0` 支持 |
+| copyable | 是否可拷贝，为对象时可进行各种自定义，说明[见下](#copyable) | boolean \| { text: string, onCopy: function, icon: ReactNode, tooltips: boolean \| ReactNode } | false | icon and tooltips: 4.4.0 |
 | delete | 添加删除线样式 | boolean | false |  |
 | disabled | 禁用文本 | boolean | false |  |
-| editable | 是否可编辑，为对象时可对编辑进行控制 | boolean \| { editing: boolean, maxLength: number, autoSize: true \| false \| { minRows: number, maxRows: number }, onStart: function, onChange: function(string) } | false | maxLength and autoSize: 4.6.0 |
+| editable | 是否可编辑，为对象时可对编辑进行控制 | boolean \| { editing: boolean, maxLength: number, autoSize: true \| false \| { minRows: number, maxRows: number }, onStart: function, onChange: function(string), icon: ReactNode, tooltip: boolean \| ReactNode } | false | maxLength, autoSize, icon and tooltip: 4.6.0 |
```
Comment (afc163) [inline components/typography/index.zh-CN.md:59]: 原来的版本信息链到下面去。

Comment (afc163) [reply components/typography/index.zh-CN.md:59]: editable 也一样处理。

**Review by afc163 — APPROVED** [review]

**Comment by afc163** [comment]
https://github.com/ant-design/ant-design/pull/26173#issuecomment-673207784

这个需要支持一下。

**Comment by llwslc** [comment]
> [#26173 (comment)](https://github.com/ant-design/ant-design/pull/26173#issuecomment-673207784)
> 
> 这个需要支持一下。

这样用就可以支持

```html
<Paragraph
  copyable={{
    icon: [
      <Tooltip title="copy tooltip" key="copy">
        <SmileOutlined />
      </Tooltip>,
      <Tooltip title="copied tooltip" key="copied">
        <SmileFilled />
      </Tooltip>,
    ],
    tooltips: false,
  }}
>
  Custom tooltip.
</Paragraph>
```

**Comment by afc163** [comment]
`tooltips: ['', '']`

这种也需要支持。

包括

`tooltips: ['', 'xxx']` 和 `tooltips: ['xxx', '']`

**Pull request merged**